Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,167 @@
+/-
+Copyright (c) 2024 María Inés de Frutos Fernández. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: María Inés de Frutos Fernández, Xavier Généreux
+-/
+import Mathlib.Data.Finsupp.Basic
+
+/-!
+This file presents a skewed version of `Mathlib.Algebra.MonoidAlgebra.Basic` with an
+irreducible definition.
+
+The definition is separated from `Finsupp` by wrapping it with a structure.
+For #15878, the goal is only to get this separation right. This means that
+most of what makes these objects skewed is currently missing from this PR.
+
+The goal will then be to define a skewed convolution product on `SkewMonoidAlgebra k G`.
+Here, the product of two elements `f g : SkewMonoidAlgebra k G` is the finitely supported
+function whose value at `a` is the sum of `f x * (x • g y)` over all pairs `x, y`
+such that `x * y = a`. (See #10541 at line 558 for an implementation.)
+
+The associativity of the skewed multiplication depends on the `[MulSemiringAction G k]` instance.
+In particular, this means that unlike in `Mathlib.Algebra.MonoidAlgebra.Basic`, `G` will
+need to be a monoid for most of our uses.
+-/
+
+noncomputable section
+
+/-- The skew monoid algebra over a semiring `k` generated by the monoid `G`.
+It is the type of finite formal `k`-linear combinations of terms of `G`,
+endowed with a skewed convolution product. -/
+structure SkewMonoidAlgebra (k : Type*) (G : Type*) [Zero k] where
+  /-- Map **from** `G →₀ k`. -/
+  ofFinsupp ::
+  /-- Map **to** `G →₀ k`. -/
+  toFinsupp : G →₀ k
+
+open Function hiding Commute
+namespace SkewMonoidAlgebra
+
+universe u₁ u₂
+
+variable {k : Type u₁} {G : Type u₂} {H : Type*} {R : Type*}
+
+section AddCommMonoid
+
+variable [AddCommMonoid k]
+
+@[simp]
+theorem eta (f : SkewMonoidAlgebra k G) : ofFinsupp f.toFinsupp = f := rfl
+
+@[irreducible]
+private def add :
```

Comment:
Why is this (and `smul`) `private`? And why are they irreducible?

Reply:
I think the idea is that we eventually don't want to be able to reference these definitions unless we really need to. The point of the separation is to forget about the `Finsupp` structure, so we ultimately want to be in a position where we never need to unfold these. That being said, we need to use it a couple time to get there.

This is my understanding of what the separation we are doing means but I could be wrong as to my knowledge there are no clear guidelines.

Reply:
They are private to hide `Finsupp` (as is done in the current implementation of `Polynomial`); same for the irreducibility of `add`.

I have removed the `irreducible` tag from `smul` since, as @eric-wieser pointed out [here](https://github.com/leanprover-community/mathlib4/pull/10541#discussion_r1705688680), this causes diamonds. In particular, the following example fails if `smul` is irreducible.
```lean
example {S} [Monoid S] [AddCommMonoid k] [DistribMulAction S k] :
    (Units.instDistribMulAction : DistribMulAction Sˣ (SkewMonoidAlgebra k G)) =
    SkewMonoidAlgebra.distribMulAction := by
  with_reducible_and_instances rfl
```